### PR TITLE
Spark on Ray on Vertex AI notebook

### PR DESCRIPTION
**Pull request by ravi-dalal**
**REQUIRED:** Add a summary of your PR here, typically including why the change is needed and what was changed. Include any design alternatives for discussion purposes.

<br>
This notebook demonstrates how Spark applications can be run on Ray on Vertex AI through a few examples.
<br><br><br>

**REQUIRED:** Fill out the below checklists or remove if irrelevant
1. If you are opening a PR for `Official Notebooks` under the [notebooks/official](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/official) folder, follow this mandatory checklist:
- [x] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/notebook_template.ipynb) as a starting point.
- [x] Follow the style and grammar rules outlined in the above notebook template.
- [x] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [x] Passes all the required automated checks. You can locally test for formatting and linting with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).
- [x] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [x] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/CODEOWNERS) file under the `Official Notebooks` section, pointing to the author or the author's team.
- [x] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

<br>

2. If you are opening a PR for `Community Notebooks` under the [notebooks/community](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/community) folder:
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/CODEOWNERS) file under the `Community Notebooks` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

<br>

3. If you are opening a PR for `Community Content` under the [community-content](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/community-content) folder:
- [ ] Make sure your main `Content Directory Name` is descriptive, informative, and includes some of the key products and attributes of your content, so that it is differentiable from other content
- [ ] The main content directory has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/community-content/CODEOWNERS) file under the `Community Content` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

### CODE REVIEW

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "tvgnzT1CKxrO"
+      },
+      "source": [
+        "## Overview\n",
+        "\n",
+        "This tutorial demonstrate how Spark can be run on Ray on Vertex AI using [RayDP](https://github.com/oap-project/raydp).\n",
+        "\n",
+        "Learn more about [Ray on Vertex AI overview](https://cloud.google.com/vertex-ai/docs/open-source/ray-on-vertex-ai/overview)."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:79]: remove "overview":
- Ray on Vertex AI

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 2]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "b55a6be7e27e"
+      },
+      "source": [
+        "### Objective\n",
+        "\n",
+        "In this tutorial, you learn how to use RayDP to run Spark applications on Ray on Vertex AI cluster.\n",
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:90]: - change from: "on Ray on Vertex AI cluster"
- to: "on a Ray cluster on Vertex AI"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 3]
```diff
@@ -0,0 +1,1353 @@
...
+        "\n",
+        "- Ray on Vertex AI\n",
+        "- Artifact Registry\n",
+        "- Cloud Storage\n",
+        "\n",
+        "\n",
+        "The steps performed include:\n",
+        "\n",
+        "- Create custom Ray on Vertex AI container image\n",
+        "- Create a Ray on Vertex AI cluster using custom container image\n",
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:102]: change: Ray on Vertex AI cluster
to: Ray cluster on Vertex AI

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 4]
```diff
@@ -0,0 +1,1353 @@
...
+        "- Run Spark interactively on the cluster using RayDP\n",
+        "- Run Spark application on cluster via Ray Job API\n",
+        "- Read files from Google Cloud Storage in Spark application\n",
+        "- Pandas UDF in Spark application on Ray on Vertex AI\n",
+        "- Delete Ray on Vertex AI cluster"
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:107]: - change: Ray on Vertex AI cluster
- to: Ray cluster on Vertex AI

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 5]
```diff
@@ -0,0 +1,1353 @@
...
+        "## Create custom container image"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "53100b3cbe10"
+      },
+      "source": [
+        "[Ray on Vertex AI container images](https://cloud.google.com/vertex-ai/docs/supported-frameworks-list#ray) do not come with RayDP pre-installed, Create a custom Ray on Vertex AI container image to run Spark applications on Ray on Vertex AI. The following section explains how a custom container image for Ray on Vertex AI with RayDP can be built."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:283]: - change: "do not"
- to: "don't"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 6]
```diff
@@ -0,0 +1,1353 @@
...
+        "!mkdir -p {DOCKER_DIR}"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "b0a46bce3de8"
+      },
+      "source": [
+        "Recommended to use the latest Ray on Vertex AI prebuilt image for creating the custom container image. Install other Python packages that are expected to be used by the Spark applications. \n",
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:323]: - change: "Recommended to use the latest Ray on Vertex AI prebuilt image..."
- to: "Recommended using the latest Ray on Vertex AI prebuilt image..."

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 7]
```diff
@@ -0,0 +1,1353 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "b37f718f7eae"
+      },
+      "source": [
+        "### Create a private Docker repository\n",
+        "Create a Docker repository in Google Artifact Registry."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:372]: - remove "Google"
- add link to the Artifact Registry overview page:
    - [Artifact Registry](https://cloud.google.com/artifact-registry/docs/overview)

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 8]
```diff
@@ -0,0 +1,1353 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "93ef2bf0de2d"
+      },
+      "source": [
+        "### Build container image\n",
+        "This tutorial assumes that [Docker](https://docs.docker.com/engine/install/) is installed and available in the work environment."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:405]: - change: "assumes"
- to: "requires"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 9]
```diff
@@ -0,0 +1,1353 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "57ad7b911d1d"
+      },
+      "source": [
+        "## Create Ray cluster on Vertex AI\n",
+        "Use the custom container image to create Ray cluster on Vertex AI using Vertex AI Python SDK."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:447]: - insert "a": "Use the custom container image to create a Ray cluster on Vertex AI using Vertex AI Python SDK"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 10]
```diff
@@ -0,0 +1,1353 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "de35e5d0026b"
+      },
+      "source": [
+        "## Spark on Ray on Vertex AI using Ray client\n",
+        "Ray [Task](https://docs.ray.io/en/latest/ray-core/tasks.html#ray-remote-functions) or [Actor](https://docs.ray.io/en/latest/ray-core/actors.html), is required for creating a Spark session with [Ray client](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/ray-client.html) on the Ray on Vertex AI.  The following code shows how a Ray Actor can be used for creating a Spark Session, running a Spark application and stopping a Spark session on a Ray on Vertex AI using RayDP.\n"
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:516]: - add comma after "Spark application":
    - "running a Spark application, and stopping a Spark session"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 11]
```diff
@@ -0,0 +1,1353 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "d7cd3ab8bd43"
+      },
+      "source": [
+        "## Spark on Ray on Vertex AI using Ray Job API\n",
+        "Ray client is useful for small experiments that require interactive connection with the Ray cluster, the Ray Job API is the recommended way to run long running and production jobs on a Ray cluster. This also applies to running Spark applications on the Ray cluster on Vertex AI."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:664]: add hyphen: "long-running"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 12]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "87467aad45e4"
+      },
+      "source": [
+        "### Use the Google Cloud Storage  Connector\n",
+        "\n",
+        "The [Google Cloud Connector for Hadoop](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/README.md) can be used for reading files from a GCS bucket from a Spark application running on Ray on Vertex AI. This is done using a few configuration parameters when a Spark session is created using RayDP. The following code shows how a CSV file stored in a GCS bucket can be read from a Spark application.\n",
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:869]: remove abbreviation "GCS"and use "Google Storage",  and add a link 
    - "reading files from a [Cloud Storage bucket](https://cloud.google.com/storage/docs/buckets) from a Spark"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 13]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "87d80b8a6a62"
+      },
+      "source": [
+        "### Use Ray data\n",
+        "\n",
+        "[Ray data API](https://docs.ray.io/en/latest/data/api/api.html) provides very convenient methods to read files from GCS bucket and it also leverages Ray's distributed processing for reading data."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:962]: - change: "GCS"
- to: "Cloud Storage"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 14]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "926bf8e334b2"
+      },
+      "source": [
+        "## Pandas UDF on Ray on Vertex AI\n",
+        "\n",
+        "The [Pyspark Pandas UDF](https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.functions.pandas_udf.html) may sometimes require additional code when they are used in a Spark application running on a Ray cluster on Vertex AI. "
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:1050]: - change: "they are"
- to: "they're"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 15]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "2148885132f1"
+      },
+      "source": [
+        "### Handle Python package dependencies\n",
+        "\n",
+        "The [Python dependencies](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/sdk.html#dependency-management) of an application can be installed using Runtime Environment with Ray job API when the Ray job is submitted to the cluster, Ray installs those dependencies in the Python virtual environment that it creates for running the job. The Pandas UDF, however, do nt run in the same python virtual environment. It instead is run in the python System environment. If that dependency is not available in the System environment, that dependency needs to be installed within Pandas UDF."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:1061]: - change: "do nt"
    - to: "don't
- change: "is not"
    - to: "isn't"

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 16]
```diff
@@ -0,0 +1,1353 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ba50e964a36a"
+      },
+      "source": [
+        "### Handle local Python dependencies\n",
+        "\n",
+        "The recommended best practice for handling Python dependencies is via Python repository. Therefore, publish your own custom packages to your Python repository and install those packages using pip. In case you are using local python package dependency in Pandas UDF of your Spark application, additional code is required to add the local packages to PYTHONPATH of Python System environment of Ray cluster on Vertex AI nodes."
```
Comment (kittyabs) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:1168]: don't need "recommended" here. It's redundant.

**Review by kittyabs — COMMENTED** [review]
Numerous edits. Note: Guidelines say "don't use abbreviations" for product/services names.

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 17]
```diff
@@ -0,0 +1,1470 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "3c482e0c7b1c"
+      },
+      "outputs": [],
+      "source": [
+        "! mkdir -p {DOCKER_DIR}"
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:311]: Can we combine this cell with the previous one?

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 18]
```diff
@@ -0,0 +1,1470 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "34d681acb474"
+      },
+      "outputs": [],
+      "source": [
+        "! gcloud artifacts repositories create {DOCKER_REPOSITORY} --repository-format=docker --location={LOCATION} --description=\"Docker repository\""
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:392]: Can we combine these two code cells as well?

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 19]
```diff
@@ -0,0 +1,1470 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "7a1451756ff8"
+      },
+      "outputs": [],
+      "source": [
+        "from google.cloud.aiplatform import vertex_ray\n",
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:493]: Can we include these imports in the same code cell where they're used below?

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 20]
```diff
@@ -0,0 +1,1470 @@
...
+      "metadata": {
+        "id": "e6a42d72e2b3"
+      },
+      "outputs": [],
+      "source": [
+        "import time\n",
+        "\n",
+        "for i in range(0, 300):\n",
+        "    time.sleep(1)\n",
+        "    print(f\"Waited for total {i} seconds...\")"
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:623]: I understand for testing purposes that we need to have this timeout here, but is there a cleaner way to implement it? Is it possible to retry the init call in the next cell multiple times if there's a failure?

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 21]
```diff
@@ -0,0 +1,1470 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "fc736ee84ef3"
+      },
+      "outputs": [],
+      "source": [
+        "! mkdir -p {SCRIPT_DIR}"
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:742]: Can we combine these two code cells?

File: notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb [hunk 22]
```diff
@@ -0,0 +1,1470 @@
...
+        "    raydp.stop_spark()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "90cb909f991b"
+      },
+      "source": [
+        "#### Submit the job"
```
Comment (katiemn) [inline notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb:1327]: Is it worth creating a helper function for this code to reuse in the steps above?

**Review by katiemn — APPROVED** [review]

**Pull request merged**